lxd/device/disk.go

```diff
@@ -552,6 +559,61 @@ func (d *disk) createDevice() (string, error) {
 	isFile := false
 	if d.config["pool"] == "" {
 		isFile = !shared.IsDir(srcPath) && !IsBlockdev(srcPath)
+		if strings.HasPrefix(d.config["source"], "cephfs") {
+			//filesystem mount
+			fields := strings.SplitN(d.config["source"], ":", 2)
+			fsName := fields[1]
+			userName := d.config["ceph.user_name"]
+			if userName == "" {
+				userName = "admin"
+			}
+			clusterName := d.config["ceph.cluster_name"]
+			if clusterName == "" {
+				clusterName = "ceph"
+			}
+			path := d.config["path"]
+			err := diskCephfsMount(clusterName, userName, fsName, path)
+			if err != nil {
+				msg := fmt.Sprintf("Could not mount Ceph FS: %s.", err)
+				if !isRequired {
+					// Will fail the PathExists test below.
+					logger.Warn(msg)
+				} else {
+					return "", fmt.Errorf(msg)
+				}
+			}
+		} else if strings.HasPrefix(d.config["source"], "ceph") {
+			// get pool name, volume name, ceph.user_name, and ceph.cluster_name from d.config and make call to map
+			// after call to map, save the src path it returned in variable src_path
+			fields := strings.SplitN(d.config["source"], ":", 2)
+			fields = strings.SplitN(fields[1], "/", 2)
+			poolName := fields[0]
+			volumeName := fields[1]
+			userName := d.config["ceph.user_name"]
+			if userName == "" {
+				userName = "admin"
+			}
+			clusterName := d.config["ceph.cluster_name"]
+			if clusterName == "" {
+				clusterName = "ceph"
+			}
+			rbdPath, err := diskCephRbdMap(clusterName, userName, poolName, volumeName)
+			if err != nil {
+				msg := fmt.Sprintf("Could not mount map Ceph RBD: %s.", err)
+				if !isRequired {
+					// Will fail the PathExists test below.
+					logger.Warn(msg)
+				} else {
+					return "", fmt.Errorf(msg)
+				}
+			}
+			err = d.volatileSet(map[string]string{"ceph_rbd_src_path": rbdPath})
+			if err != nil {
+				return "", err
+			}
```

Comment:
This one requires far less changes than the `cephfs` one I just detailed above.
It's effectively correct expect that `srcPath` must be set to the value of `rbdPath`.

This should then allow `DiskMount` to do its job and will also get you through the `PathExists` test.

Reply:
Yup I see, essentially what we had earlier was almost correct with srcPath. Instead of renaming the variable and trimming we should have just set it to the result of the rbdmap call